From 030cffaa6d5cdde32588caf28ba8be5d79be80a1 Mon Sep 17 00:00:00 2001
From: Martin Peres <martin.peres@linux.intel.com>
Date: Wed, 6 Jun 2018 17:59:02 +0300
Subject: [PATCH 02/10] modesetting: make drmmode_shadow_create more generic

To do so:
 - add a pixmap field to the drmmode_scanout_rec structure
 - split out the pixmap-creation out of drmmode_shadow_create
   into drmmode_scanout_pixmap_create()

Signed-off-by: Martin Peres <martin.peres@linux.intel.com>
---
 .../drivers/modesetting/drmmode_display.c     | 71 +++++++++++--------
 .../drivers/modesetting/drmmode_display.h     |  1 +
 2 files changed, 44 insertions(+), 28 deletions(-)

diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index cba45775d..886115ca1 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -1878,50 +1878,65 @@ drmmode_create_pixmap_header(ScreenPtr pScreen, int width, int height,
 static Bool
 drmmode_set_pixmap_bo(drmmode_ptr drmmode, PixmapPtr pixmap, drmmode_bo *bo);
 
-static PixmapPtr
-drmmode_shadow_create(xf86CrtcPtr crtc, void *data, int width, int height)
+static Bool
+drmmode_scanout_pixmap_create(xf86CrtcPtr crtc, drmmode_shadow_scanout_ptr scanout)
 {
     ScrnInfoPtr scrn = crtc->scrn;
     drmmode_crtc_private_ptr drmmode_crtc = crtc->driver_private;
     drmmode_ptr drmmode = drmmode_crtc->drmmode;
-    uint32_t rotate_pitch;
-    PixmapPtr rotate_pixmap;
+    uint32_t pitch;
+    PixmapPtr pixmap;
     void *pPixData = NULL;
 
-    if (!data) {
-        data = drmmode_shadow_allocate(crtc, width, height);
-        if (!data) {
-            xf86DrvMsg(scrn->scrnIndex, X_ERROR,
-                       "Couldn't allocate shadow pixmap for rotated CRTC\n");
-            return NULL;
-        }
-    }
-
-    if (!drmmode_bo_has_bo(&drmmode_crtc->shadow_rotated.bo)) {
+    if (!drmmode_bo_has_bo(&scanout->bo)) {
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
-                   "Couldn't allocate shadow pixmap for rotated CRTC\n");
-        return NULL;
+                   "Couldn't allocate scanout pixmap because no scanout BO is present\n");
+        return FALSE;
     }
 
-    pPixData = drmmode_bo_map(drmmode, &drmmode_crtc->shadow_rotated.bo);
-    rotate_pitch = drmmode_bo_get_pitch(&drmmode_crtc->shadow_rotated.bo),
+    pPixData = drmmode_bo_map(drmmode, &scanout->bo);
+    pitch = drmmode_bo_get_pitch(&scanout->bo),
 
-    rotate_pixmap = drmmode_create_pixmap_header(scrn->pScreen,
-                                                 width, height,
-                                                 scrn->depth,
-                                                 drmmode->kbpp,
-                                                 rotate_pitch,
-                                                 pPixData);
+    pixmap = drmmode_create_pixmap_header(scrn->pScreen,
+                                          scanout->bo.width,
+                                          scanout->bo.height,
+                                          scrn->depth,
+                                          drmmode->kbpp,
+                                          pitch,
+                                          pPixData);
 
-    if (rotate_pixmap == NULL) {
+    if (pixmap == NULL) {
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "Couldn't allocate shadow pixmap for rotated CRTC\n");
-        return NULL;
+        return FALSE;
     }
 
-    drmmode_set_pixmap_bo(drmmode, rotate_pixmap, &drmmode_crtc->shadow_rotated.bo);
+    if (!drmmode_set_pixmap_bo(drmmode, pixmap, &scanout->bo))
+        return FALSE;
 
-    return rotate_pixmap;
+    scanout->pixmap = pixmap;
+    return TRUE;
+}
+
+static PixmapPtr
+drmmode_shadow_create(xf86CrtcPtr crtc, void *data, int width, int height)
+{
+    ScrnInfoPtr scrn = crtc->scrn;
+    drmmode_crtc_private_ptr drmmode_crtc = crtc->driver_private;
+
+    if (!data) {
+        data = drmmode_shadow_allocate(crtc, width, height);
+        if (!data) {
+            xf86DrvMsg(scrn->scrnIndex, X_ERROR,
+                       "Couldn't allocate shadow pixmap for rotated CRTC\n");
+            return NULL;
+        }
+    }
+
+    if (drmmode_scanout_pixmap_create(crtc, &drmmode_crtc->shadow_rotated))
+        return drmmode_crtc->shadow_rotated.pixmap;
+    else
+        return NULL;
 }
 
 static void
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.h b/hw/xfree86/drivers/modesetting/drmmode_display.h
index 6f3643f3e..b7e32e272 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.h
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.h
@@ -156,6 +156,7 @@ typedef struct {
 typedef struct {
     drmmode_bo bo;
     unsigned fb_id;
+    PixmapPtr pixmap;
 } drmmode_shadow_scanout_rec, *drmmode_shadow_scanout_ptr;
 
 typedef struct {
-- 
2.25.1

